streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -0,0 +1,258 @@
+            consumer.assign(Collections.singletonList(topicPartition));
+            final Long checkpoint = checkpointableOffsets.get(topicPartition);
+            if (checkpoint != null) {
+                consumer.seek(topicPartition, checkpoint);
+            } else {
+                consumer.seekToBeginning(Collections.singletonList(topicPartition));
+            }
+
+            long offset = consumer.position(topicPartition);
+            final Long highWatermark = highWatermarks.get(topicPartition.partition());
+
+            while (offset < highWatermark) {
+                final ConsumerRecords<byte[], byte[]> records = consumer.poll(100);
+                for (ConsumerRecord<byte[], byte[]> record : records) {
+                    offset = record.offset() + 1;
+                    stateRestoreCallback.restore(record.key(), record.value());
+                }
+            }
+            checkpointableOffsets.put(topicPartition, offset);
+        }
+    }
+
+    public void flush(final InternalProcessorContext context) {
+        if (!this.stores.isEmpty()) {
+            log.debug("Flushing all global stores registered in the state manager");
+            for (StateStore store : this.stores.values()) {
+                final ProcessorNode processorNode = topology.storeToProcessorNodeMap().get(store);
+                if (processorNode != null) {
+                    context.setCurrentNode(processorNode);
+                }
+                try {
+                    log.trace("Flushing global store={}", store.name());
+                    store.flush();
+                } catch (Exception e) {
+                    throw new ProcessorStateException(String.format("Failed to flush global state store %s", store.name()), e);
+                }
+            }
+        }
+    }
+
+
+    @Override
+    public void close(final Map<TopicPartition, Long> offsets) throws IOException {
+        try {
+            if (stores.isEmpty()) {
+                return;
+            }
+            final StringBuilder closeFailed = new StringBuilder();
+            for (final Map.Entry<String, StateStore> entry : stores.entrySet()) {
+                log.debug("Closing global storage engine {}", entry.getKey());
+                try {
+                    entry.getValue().close();
+                } catch (Exception e) {
+                    log.error("Failed to close global state store {}", entry.getKey(), e);
+                    closeFailed.append("Failed to close global state store:")
+                            .append(entry.getKey())
+                            .append(". Reason: ")
+                            .append(e.getMessage())
+                            .append("\n");
+                }
+            }
+            stores.clear();
+            if (closeFailed.length() > 0) {
+                throw new ProcessorStateException("Exceptions caught during close of 1 or more global state stores\n" + closeFailed);
```

Comment:
It probably should be deleted after it is first loaded. At least that is what we do elsewhere.